pkg/k8shandler/status.go

```diff
@@ -261,147 +264,183 @@ func (er *ElasticsearchRequest) updateNodeConditions(status *api.ElasticsearchSt
-					if containerStatus.State.Waiting != nil {
-						updatePodNotReadyCondition(
-							node,
-							api.ESContainerWaiting,
-							containerStatus.State.Waiting.Reason,
-							containerStatus.State.Waiting.Message,
-						)
-					} else {
-						updatePodNotReadyCondition(
-							node,
-							api.ESContainerWaiting,
-							"",
-							"",
-						)
-					}
-					if containerStatus.State.Terminated != nil {
-						updatePodNotReadyCondition(
-							node,
-							api.ESContainerTerminated,
-							containerStatus.State.Terminated.Reason,
-							containerStatus.State.Terminated.Message,
-						)
-					} else {
-						updatePodNotReadyCondition(
-							node,
-							api.ESContainerTerminated,
-							"",
-							"",
-						)
-					}
-				}
-				if containerStatus.Name == "proxy" {
-					if containerStatus.State.Waiting != nil {
-						updatePodNotReadyCondition(
-							node,
-							api.ProxyContainerWaiting,
-							containerStatus.State.Waiting.Reason,
-							containerStatus.State.Waiting.Message,
-						)
-					} else {
-						updatePodNotReadyCondition(
-							node,
-							api.ProxyContainerWaiting,
-							"",
-							"",
-						)
-					}
-					if containerStatus.State.Terminated != nil {
-						updatePodNotReadyCondition(
-							node,
-							api.ProxyContainerTerminated,
-							containerStatus.State.Terminated.Reason,
-							containerStatus.State.Terminated.Message,
-						)
-					} else {
-						updatePodNotReadyCondition(
-							node,
-							api.ProxyContainerTerminated,
-							"",
-							"",
-						)
+		er.pruneStatus(status, nodeIndex, nodeName, nodePodList.Items)
+		er.updatePodStateMap(nodeName, nodeStatus, nodePodList.Items, thresholdEnabled)
+	}
+
+	return nil
+}
+
+func (er *ElasticsearchRequest) pruneStatus(status *api.ElasticsearchStatus, nodeIndex int, nodeName string, pods []v1.Pod) {
+	clusterNodes := nodes[nodeMapKey(er.cluster.GetName(), er.cluster.GetNamespace())]
+	nodeStatus := status.Nodes[nodeIndex]
+	ll := er.L().WithValues("node", nodeStatus)
+
+	for _, node := range clusterNodes {
+		if nodeName == node.name() && node.isMissing() {
+			ll.Info("Pruning node status")
+			// Prune node status list for non existing nodes
+			status.Nodes = append(status.Nodes[:nodeIndex], status.Nodes[nodeIndex+1:]...)
+
+			if len(pods) == 0 {
+				for nodeRole, podStateMap := range status.Pods {
+					for podState, podNames := range podStateMap {
+						for idx, podName := range podNames {
+							if strings.HasPrefix(podName, nodeName) {
+								// Prune pod state maps for non existing pods
+								status.Pods[nodeRole][podState] = append(podNames[:idx], podNames[idx+1:]...)
+							}
+						}
 					}
 				}
 			}
+		}
+	}
+}
 
-			if !thresholdEnabled {
-				// disk threshold is not enabled, continue to next node
-				continue
-			}
+func (er *ElasticsearchRequest) updatePodStateMap(nodeName string, nodeStatus *api.ElasticsearchNodeStatus, pods []v1.Pod, thresholdEnabled bool) {
```

Comment:
this function name is a little misleading since it does more than just update the pod state map, it also updates the node conditions

Reply:
Renamed to `updatePodNodeConditions`. WDYT?